sdk/servicebus/Azure.Messaging.ServiceBus/src/Diagnostics/ServiceBusEventSource.cs

```diff
@@ -1509,7 +1592,34 @@ public void TransactionDischarged(string localTransactionId, ArraySegment<byte>
         [Event(TransactionDischargedEvent, Level = EventLevel.Informational, Message = "AmqpTransactionDischarged for LocalTransactionId: {0} AmqpTransactionId: {1} Rollback: {2}.")]
         public void TransactionDischarged(string transactionId, string amqpTransactionId, bool rollback)
         {
-            WriteEvent(TransactionDischargedEvent, transactionId, amqpTransactionId, rollback);
+            if (IsEnabled())
+            {
+                TransactionDischargedCore(TransactionDischargedEvent, transactionId, amqpTransactionId, rollback);
+            }
+        }
+
+        [NonEvent]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private unsafe void TransactionDischargedCore(int eventId, string transactionId, string amqpTransactionId, bool rollback)
+        {
+            fixed (char* transactionIdPtr = transactionId)
+            fixed (char* amqpTransactionIdPtr = amqpTransactionId)
+            {
+                var eventPayload = stackalloc EventData[3];
+
+                eventPayload[0].Size = (transactionId.Length + 1) * sizeof(char);
+                eventPayload[0].DataPointer = (IntPtr)transactionIdPtr;
+
+                eventPayload[1].Size = (amqpTransactionId.Length + 1) * sizeof(char);
+                eventPayload[1].DataPointer = (IntPtr)amqpTransactionIdPtr;
+
+                // bool maps to "win:Boolean", a 4-byte boolean
```

Comment:
Interesting; I'm assuming this has something to do with how we were marshalling the boolean and the expectations downstream.  Can you help me understand a bit more context?

Reply:
My understanding is that due to alignment a bool needs up to four bytes. But since I used `Unsafe.SizeOf<bool>` as the size, which returns 1for the size, but the pointer points to a bool that gets aligned to four bytes this causes problem when reading it. This then crashes with an argument exception